File: backend/app/api/jobs.py
"""The public job feed."""

from datetime import datetime
from typing import Annotated

from fastapi import APIRouter, Depends, Header, HTTPException, Query, status
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.db import get_session
from app.domain.role_family import ROLE_FAMILIES, Seniority
from app.domain.scoring import MatchBreakdown
from app.models.student import Student
from app.security import read_access_token
from app.services import job_service
from app.services.job_service import JobFilters
from app.services.scoring_service import (
    MAX_RANKED,
    get_student_profile,
    rank_by_score,
    score_page,
)

router = APIRouter(prefix="/jobs", tags=["jobs"])

_SENIORITIES = {str(s) for s in Seniority}
_COUNTRIES = {"US", "CA"}


class ScoreBreakdown(BaseModel):
    total: int
    skill_points: int
    experience_points: int
    keyword_points: int
    freshness_points: int
    skill_state: str
    experience_state: str
    keyword_state: str
    freshness_state: str
    matched_skills: list[str] = []
    missing_skills: list[str] = []
    required_years: float | None = None
    requirement_basis: str | None = None
    requirement_phrase: str | None = None


class JobSummary(BaseModel):
    id: int
    source: str
    company_name: str
    title: str
    location_raw: str | None
    country: str | None
    is_remote: bool | None
    role_family: str | None
    seniority: str | None
    posted_at: datetime | None
    first_seen_at: datetime
    closed_at: datetime | None

    # Whether the company's own board carries this posting. Not decoration: an aggregator copy
    # is a weaker claim, and the interface says which it is.
    is_verified: bool

    # Present only when the request is authenticated and the student has an active resume.
    score: ScoreBreakdown | None = None


class JobAlias(BaseModel):
    """Another record of the same job, collapsed into the canonical row."""

    source: str
    apply_url: str
    is_verified: bool


class JobDetail(JobSummary):
    description: str
    apply_url: str
    # Where else this posting was seen. Shown so collapsing is visible: telling a student two
    # records are one job is a claim, and an unexplained claim has to be taken on trust.
    also_seen_on: list[JobAlias] = []


class JobFeed(BaseModel):
    items: list[JobSummary]
    total: int
    page: int
    page_size: int
    # Echoed back so an empty feed can name the filter responsible instead of only reporting
    # that nothing matched.
    applied_filters: dict[str, object]

    # Whether these items carry scores. False for an anonymous request or a student with no
    # parsed resume, so the interface can explain the absence rather than showing empty bars.
    scored: bool = False

    # How many postings were ranked. Below the total when the filtered set exceeds the ranking
    # bound, and shown rather than hidden: a student sorting by match deserves to know the
    # ordering covers the first 200 rather than everything.
    ranked_within: int | None = None


def _csv(value: str | None, allowed: set[str], name: str) -> list[str]:
    """Parse a comma-separated filter, rejecting anything unrecognised.

    Rejected rather than ignored. Silently dropping an unknown value shows the student a feed
    they did not ask for, from which they would reasonably conclude the filter does nothing.
    """
    if not value:
        return []

    parts = [part.strip() for part in value.split(",") if part.strip()]
    unknown = [part for part in parts if part not in allowed]
    if unknown:
        raise HTTPException(
            # The spelling starlette now wants; the ENTITY name is deprecated and the
            # project treats DeprecationWarning as an error.
            status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
            detail=(
                f"Unknown {name}: {', '.join(unknown)}. "
                f"Choose from {', '.join(sorted(allowed))}."
            ),
        )
    return parts


async def _optional_student(
    session: AsyncSession = Depends(get_session),
    authorization: Annotated[str | None, Header()] = None,
) -> tuple[Student | None, int | None]:
    """The authenticated student if one exists, without requiring auth.

    The feed is public. An unauthenticated request sees the index without scores; an
    authenticated one gets scores added to each item. This must never become a hard dependency,
    or browsing the index requires registration — which would make the demo useless to judges.
    """
    if not authorization:
        return None, None
    try:
        parts = authorization.split()
        if len(parts) != 2 or parts[0].lower() != "bearer":
            return None, None
        user_id = read_access_token(parts[1])
    except Exception:
        return None, None

    stmt = select(Student).where(Student.user_id == user_id)
    student = (await session.execute(stmt)).scalars().first()
    if student is None:
        return None, None

    from app.models.resume import ResumeMaster

    resume_stmt = select(ResumeMaster.id).where(
        ResumeMaster.student_id == student.id, ResumeMaster.is_active.is_(True)
    )
    resume_id = (await session.execute(resume_stmt)).scalars().first()
    return student, resume_id


@router.get("", response_model=JobFeed)
async def list_jobs(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    seniority: str | None = Query(None, description="Comma separated: entry, senior, unknown"),
    role_family: str | None = Query(None, description="Comma separated role families"),
    country: str | None = Query(None, description="Comma separated: US, CA"),
    remote: bool | None = Query(None),
    q: str | None = Query(None, max_length=200),
    company: str | None = Query(None, max_length=200),
    include_closed: bool = Query(
        False,
        description=(
            "Include postings that have closed. Off by default: a closed role is not a "
            "weaker match, it is not a match."
        ),
    ),
    session: AsyncSession = Depends(get_session),
    student_and_resume: tuple[Student | None, int | None] = Depends(_optional_student),
) -> JobFeed:
    student, resume_id = student_and_resume

    filters = JobFilters(
        seniority=_csv(seniority, _SENIORITIES, "seniority"),
        role_family=_csv(role_family, set(ROLE_FAMILIES), "role_family"),
        country=_csv(country, _COUNTRIES, "country"),
        remote=remote,
        q=q.strip() if q and q.strip() else None,
        company=company.strip() if company and company.strip() else None,
        include_closed=include_closed,
    )

    profile = await get_student_profile(session, student.id) if student and resume_id else None

    # Unscored path: no token, no resume, or a resume that has not parsed. The feed is public and
    # must work for all three, ordered by recency, so browsing the index never requires an
    # account — story 34, and the path judges see first.
    if profile is None or student is None or resume_id is None:
        result = await job_service.list_jobs(
            session, page=page, page_size=page_size, filters=filters
        )
        return JobFeed(
            items=[JobSummary.model_validate(j, from_attributes=True) for j in result.items],
            total=result.total,
            page=result.page,
            page_size=result.page_size,
            applied_filters=result.filters.as_dict(),
            scored=False,
        )

    # Scored path. Ordering by score and computing scores lazily are in tension — you cannot sort
    # by a number you have not calculated — so a bounded window of the filtered set is scored and
    # ranked. See MAX_RANKED for why the bound is where it is.
    window = await job_service.list_jobs(session, page=1, page_size=MAX_RANKED, filters=filters)
    scores = await score_page(
        session,
        student_id=student.id,
        resume_master_id=resume_id,
        jobs=window.items,
        profile=profile,
    )
    ranked = rank_by_score(window.items, scores)

    start = (page - 1) * page_size
    page_items = ranked[start : start + page_size]

    items: list[JobSummary] = []
    for job in page_items:
        summary = JobSummary.model_validate(job, from_attributes=True)
        breakdown = scores.get(job.id)
        if breakdown:
            summary.score = _to_schema(breakdown)
        items.append(summary)

    return JobFeed(
        items=items,
        total=window.total,
        page=page,
        page_size=page_size,
        applied_filters=window.filters.as_dict(),
        scored=True,
        ranked_within=min(window.total, MAX_RANKED),
    )


def _to_schema(breakdown: MatchBreakdown) -> ScoreBreakdown:
    return ScoreBreakdown(
        total=breakdown.total,
        skill_points=breakdown.skill_points,
        experience_points=breakdown.experience_points,
        keyword_points=breakdown.keyword_points,
        freshness_points=breakdown.freshness_points,
        skill_state=breakdown.skill_state.value,
        experience_state=breakdown.experience_state.value,
        keyword_state=breakdown.keyword_state.value,
        freshness_state=breakdown.freshness_state.value,
        matched_skills=breakdown.matched_skills,
        missing_skills=breakdown.missing_skills,
        required_years=breakdown.required_years,
        requirement_basis=(
            breakdown.requirement_basis.value if breakdown.requirement_basis else None
        ),
        requirement_phrase=breakdown.requirement_phrase,
    )


@router.get("/meta/filters")
async def filter_options() -> dict[str, list[str]]:
    """What the feed can be filtered by.

    Served rather than duplicated in the frontend, so a family added to the classifier becomes
    selectable without a second edit in another language.
    """
    return {
        "role_family": list(ROLE_FAMILIES),
        "seniority": sorted(_SENIORITIES),
        "country": sorted(_COUNTRIES),
    }


@router.get("/{job_id}", response_model=JobDetail)
async def get_job(
    job_id: int,
    session: AsyncSession = Depends(get_session),
    identity: tuple[Student | None, int | None] = Depends(_optional_student),
) -> JobDetail:
    job = await job_service.get_job(session, job_id)
    aliases = await job_service.get_aliases(session, job_id)

    detail = JobDetail.model_validate(job, from_attributes=True)
    detail.also_seen_on = [
        JobAlias.model_validate(alias, from_attributes=True) for alias in aliases
    ]

    # The detail page is where the score is explained: which skills matched, which the posting
    # asked for and the resume lacks, and the sentence the experience requirement was read from.
    # The feed can only afford the bar. Scoring one posting is cheap, so this is computed here
    # rather than passed through from the feed — a student who opens a job from a link, or
    # reloads the page, is owed the same explanation as one who arrived by scrolling.
    student, resume_id = identity
    if student is not None and resume_id is not None:
        profile = await get_student_profile(session, student.id)
        if profile is not None:
            scores = await score_page(
                session,
                student_id=student.id,
                resume_master_id=resume_id,
                jobs=[job],
                profile=profile,
            )
            breakdown = scores.get(job.id)
            if breakdown:
                detail.score = _to_schema(breakdown)

    return detail

File: frontend/src/lib/jobs.ts
/**
 * Reading the shared job index.
 *
 * `is_verified` is carried through to the interface rather than flattened away. A posting on
 * a company's own board and one seen only on an aggregator are different claims, and this
 * product's position is that it never presents a guess as a fact.
 */
import { apiUrl } from "./api";
import { storedToken } from "./auth";

export type ComponentState = "scored" | "met" | "short" | "unstated";

/**
 * The score, decomposed. Never rendered as a bare total — ADR 0003 and the design brief both
 * require the four parts to be visible, because a single opaque number is exactly what the
 * student already gets from every job board that ignores them.
 */
export interface ScoreBreakdown {
  total: number;
  skill_points: number;
  experience_points: number;
  keyword_points: number;
  freshness_points: number;
  skill_state: ComponentState;
  experience_state: ComponentState;
  keyword_state: ComponentState;
  freshness_state: ComponentState;
  matched_skills: string[];
  missing_skills: string[];
  required_years: number | null;
  requirement_basis: string | null;
  /** The words the requirement was read from, so the number can be checked against its source. */
  requirement_phrase: string | null;
}

export interface JobSummary {
  id: number;
  source: string;
  company_name: string;
  title: string;
  location_raw: string | null;
  country: string | null;
  is_remote: boolean | null;
  role_family: string | null;
  seniority: string | null;
  posted_at: string | null;
  first_seen_at: string;
  closed_at: string | null;
  is_verified: boolean;
  /** Absent for an anonymous request or a student with no parsed resume. */
  score: ScoreBreakdown | null;
}

export interface JobAlias {
  source: string;
  apply_url: string;
  is_verified: boolean;
}

export interface JobDetail extends JobSummary {
  description: string;
  apply_url: string;
  /** Other listings collapsed into this one, so the match is visible rather than assumed. */
  also_seen_on: JobAlias[];
}

export interface JobFeed {
  items: JobSummary[];
  total: number;
  page: number;
  page_size: number;
  /** False for an anonymous request or a student with no parsed resume. */
  scored: boolean;
  /** How many postings the ranking covered, when that is fewer than the total. */
  ranked_within: number | null;
}

export interface BoardStatus {
  provider: string;
  token: string;
  company_name: string;
  active: boolean;
  last_fetched_at: string | null;
  last_succeeded_at: string | null;
  consecutive_failures: number;
  last_error: string | null;
}

async function getJson<T>(path: string): Promise<T> {
  // The token is sent when present, and its absence is not an error. The index is public: an
  // anonymous request gets the feed unscored, a signed-in one gets scores added. Making this a
  // hard requirement would mean browsing the index required an account.
  const token = storedToken();
  const response = await fetch(apiUrl(path), {
    headers: {
      Accept: "application/json",
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
  });

  if (!response.ok) {
    // The backend envelope is {error: {code, message}}. Preferring its message means the
    // interface explains the actual problem rather than showing a status code.
    let message = `The request failed (${response.status}).`;
    try {
      const body = await response.json();
      if (body?.error?.message) message = body.error.message;
    } catch {
      // Not every failure returns JSON. A cold start or a proxy error will not.
    }
    throw new Error(message);
  }

  return (await response.json()) as T;
}

export interface JobQuery {
  seniority?: string[];
  roleFamily?: string[];
  country?: string[];
  remote?: boolean;
  q?: string;
}

export function fetchJobs(
  params: { page?: number; pageSize?: number } & JobQuery = {},
) {
  const query = new URLSearchParams();
  query.set("page", String(params.page ?? 1));
  query.set("page_size", String(params.pageSize ?? 20));
  if (params.seniority?.length) query.set("seniority", params.seniority.join(","));
  if (params.roleFamily?.length) query.set("role_family", params.roleFamily.join(","));
  if (params.country?.length) query.set("country", params.country.join(","));
  if (params.remote !== undefined) query.set("remote", String(params.remote));
  if (params.q) query.set("q", params.q);
  return getJson<JobFeed>(`/api/v1/jobs?${query.toString()}`);
}

export function fetchJob(id: number) {
  return getJson<JobDetail>(`/api/v1/jobs/${id}`);
}

export function fetchSources() {
  return getJson<{ boards: BoardStatus[] }>("/api/v1/sources");
}

export const queryKeys = {
  jobs: (page: number) => ["jobs", page] as const,
  job: (id: number) => ["job", id] as const,
  sources: () => ["sources"] as const,
};
